### ISSUE TITLE
how to get the input from the input nodes

### ISSUE THREAD

**Issue by jhubers** [DynamoDS/Dynamo#4776]
Hi,
I am trying to develop some custom nodes.
I would like to add semicolons to several inputs and then output it as one string. How do I get the input from the input nodes?

**Comment by holyjewsus** [comment]
Hi @jhubers I'm a bit confused - do you mean custom nodes developed through the UI? or do you mean nodes developed in c#?

**Comment by jhubers** [comment]
Hi,
I try developing in C#, folowing the HelloDynamo sample.

Verzonden met mijn Windows Phone
________________________________
Van: Michael Kirschner<mailto:notifications@github.com>
Verzonden: ‎23-‎6-‎2015 20:25
Aan: DynamoDS/Dynamo<mailto:Dynamo@noreply.github.com>
CC: Hans Hubers - BK<mailto:J.C.Hubers@tudelft.nl>
Onderwerp: Re: [Dynamo] how to get the input from the input nodes (#4776)


Hi @jhubers<https://github.com/jhubers> I'm a bit confused - do you mean custom nodes developed through the UI? or do you mean nodes developed in c#?

**Comment by ke-yu** [comment]
@jhubers , you could try `String.Join`:

![untitled](https://cloud.githubusercontent.com/assets/2600379/8325116/72094e60-1a89-11e5-95c6-f5b8374e373c.png)


BTW, seems we can't write `";"` in code block node. @monikaprabhu , we need to log a bug for it).

**Comment by jhubers** [comment]
@ke-yu, I mean inside a C# custom node. E.g. I have: InPortData.Add(new PortData("P", "Parameter name as string.")); Now inside my custom node I want to add ";" to what ever was connected to this InPort. Do I have to use TryGetValue. If so with what attributes?

**Comment by ke-yu** [comment]
Hi @jhubers , sorry, I thought you asked how to do that with Dynamo's custom node. 

Do you mean `CustomNodeModel` in HelloDyanmo sample?  As Dynamo internally compiles nodes to DesignScript code,  instead of directly getting value from input, we have to use DesignScript AST to implement the custom node model logic. For example, in custom node mode's `BuildOutputAst()` function: 

```
var semiColon = AstFactor.BuildStringNode(";");
var arguments = new List<AssociativeNode>() {
    // suppose custom node model only has one input
    inputAstNodes[0], 
    semiColon};

// To append ";" to input string, use +. For example, "hello" + ";"
// + operation is actually a function call %add(x, y), so we create a function call node 
var funcNode = AstFactory.BuildFunctionCall("%add", arguments);
return new[] 
{
    AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode);
};
```

You may take a look at `CustomNodeModel.BuildOutputAst()`

**Comment by jhubers** [comment]
Hi @ke-yu, It is confusing. I try to develop a dll that, when put in the Dynamo nodes folder, makes a menu entry with my nodes. One of the nodes should prepare additional input for a csv file. In the Netherlands we use semicolons as separator. I would expect that the inputs of my node are stored in a list or dictionary or  whatever kind of array. Then iterating through that I could add the semicolon to List<string> myList, to be used in AstFactory.BuildAssignment( GetAstIdentifierForOutputIndex(0), AstFactory.BuildExprList(myList)), I was not familiar with TryGetValue and thought it was a method from Dynamo, but now I found that it is a regular C# method and found the description. I will try to use that. Seems more logical to me. But otherwise use your solution. Thanks.

**Comment by ke-yu** [comment]
Hi @jhubers , if your implementation is in a C# dll, you don't need to deal with `GetAstIdentifierForOutputIndex(0)` or whatever AST node, or even don't need to deal with `InputPort` (Dynamo will automatically create for you when you import it into Dynamo), just write a normal C# function, and import C# dll into Dynamo (Edit | Import Library), then you nodes will be displayed in library view.

**Comment by jhubers** [comment]
I am trying the following now, but get an error. BTW, using Dynamo 0.8.
So in a custom node the inputs are not directly accessible as strings or other types. They are  of type AssociativeNode and stored in a List< AssociativeNode > inputAstNodes. To add a semicolon you have to convert it into Abstract Syntax Tree (AST) nodes. For this you use AstFactory: 
var semiColon = AstFactory.BuildStringNode(";"); And then build a new List< AssociativeNode > as follows:
List<AssociativeNode> pCOLLECToutputList = new List<AssociativeNode>();
foreach (AssociativeNode InputItem in inputAstNodes)
  {
      pCOLLECToutputList.Add(InputItem);
      pCOLLECToutputList.Add(semiColon);
  }
  //remove the last semiColon
  if (pCOLLECToutputList.Count > 0)
  {
      pCOLLECToutputList.RemoveAt(pCOLLECToutputList.Count - 1);
  }
However then we would get an output with the semicolon on every second line. Therefore we need a function that adds the AssociativeNodes together:
var funcNode = AstFactory.BuildFunctionCall("%add", pCOLLECToutputList);
And then we can prepare the output as follows:
return new[]
  {AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)};
However it gives the error: Warning: Method '%add()' not found.

**Comment by tatlin** [comment]
This is overkill, you do not need to deal with AST unless you need custom UI of some sort or need access to a lower-level part of dynamo. Take a look at some of the samples and tutorials for loading dlls via dynamo's "zero touch import" capability. 

https://github.com/DynamoDS/Dynamo/wiki/Zero-Touch-Plugin-Development

"In most cases, C# static methods and Classes can be imported without modification. If your library only needs to call functions, and not construct new objects, this can be achieved very easily with static methods. When Dynamo loads your DLL, it will strip off the namespace of your classes, and expose all static methods as nodes. "

Also if all you are needing is csv export via a different delimiter you should not need to write your own dll at all. Take a look at the csv exporter nodes and the excel interop nodes

-matt

**Comment by jhubers** [comment]
Hi Matt,
It is about several nodes. Preparing a string of seperated inputs to add to a csv file is just one of the functions. One of the nodes has indeed a customized UI.

**Comment by holyjewsus** [comment]
hi @jhubers did @ke-yu 's example solve the issue / get you started?

**Comment by jhubers** [comment]
hi @holyjewsus  yes I think I can get what I want. Thanks.

**Comment by jhubers** [comment]
hi @holyjewsus in fact it helped only for one node. An other node needs to add the inported data to a DataTable. Is TryGetValue then the right way to go?

**Comment by jhubers** [comment]
Found this solution. Is this the way to go?
            for (int i = 0; i < Inputs.Count ; i++)
            {
                var item = Inputs[i];
                Dynamo.Nodes.CodeBlockNodeModel itemValue = (Dynamo.Nodes.CodeBlockNodeModel)item.Item2;
                string s = itemValue.Code;
                //for some reason Dynamo puts "\" and \";" around the string
                string sCleaned = s.Remove(s.Length - 2).Remove(0, 1);
            }

**Comment by holyjewsus** [comment]
Hi @jhubers it looks like you are directly interacting with nodes in the graph, which could potentially be unsafe. It's difficult to reason about the state all other nodes are in.

What are you trying to accomplish with this work? What do you mean by a data table? If you need to access the raw code of the code block then what you're doing is probably okay, but inside your Abstract syntax tree creation you can call arbitrary c# functions you write, so you could take the inputs nodes and call your c# which writes them out to whatever data structure you need. This c# code must be defined in another assembly than the node itself it written in.

**Comment by jhubers** [comment]
Hi @holyjewsus 
First I have to state that I am not a programmer. I picked up some knowledge by doing. I am working on a research project called pCOLAD, as presented and published at the eCAADe 2014 conference. I want to extend it from Grasshopper to Dynamo. That meant from vb.net and Windows Forms to C#, xaml and MVVM (quite a step!). 

pCOLAD is about multidisciplinary parametric design. The idea is that different disciplines use some crucial shared parameters and are immediately informed if a value of a shared parameter is changed. Since a CSV file is easy for all disciplines to handle, I need a pCOLLECT node to gather the crucial parameters and values in the Dynamo solution. And link that to a pSHARE node, that displays a copy of the CSV file (a xaml control with a DataGrid bound to a DataTable) and adds the output of pCOLLECT. A pPARAM node uses the output of pSHARE to find the parameters and values that are needed from other team members (see attached image). Of course it is more complex than this. In order to understand what a parameter is about, you have to add several attributes. This must be flexible. So pCOLLECT should have the possibility to add parameter inputs. And some smart mapping of different attributes is needed. 

As written above I found a way to turn the input of pCOLLECT into CSV format output. Now I linked that to an input of pSHARE, but again have difficulties to use it (in the image it is faked). I use the Inputs Dictionary to find the data. But need to know the type of course. It turns out to be of type pCOLLECT. I was expecting a Dynamo type like Dynamo.Nodes.CodeBlockNodeModel. 

But you say this is an unsafe way. Do you mean that it is not reliable, or that in future versions of Dynamo it will not work (as I noticed going from 0.7 to 0.8). The last is not a big problem. It is a research project. Once Dynamo is more stable I plan to find funding for a test case project.
![image](https://cloud.githubusercontent.com/assets/1529297/8374915/5c6f7e32-1bf9-11e5-9494-909f72f42d21.png)

**Comment by jhubers** [comment]
Now if I connect a File Path node and check the type in Inputs, it turns out to be a DSCore.File.Filename. But this is a static type, and I cannot declare a variable of that type. How do I get the file path in my custom UI node?

**Comment by ke-yu** [comment]
hi @jhubers , `DSCore.File.Filename` is the type of `File Path` node, but `File Path` node should only output string. 

I'm afraid you have to write a little bit AST nodes, similar to `SampleLibraryUI.Examples.CustomNodeModel.BuildOutputAst()`, where

```
[IsVisibleInDynamoLibrary(false)]
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
{
    ....        
    var funcNode = AstFactory.BuildFunctionCall(
                new Func<double,double>(SampleUtilities.MultiplyInputByNumber), 
                new List<AssociativeNode>(){doubleNode});
    ....
}
```

You may implement your logic in a static class, for example:
```
public static class MyDataCollector
{
     [IsVisibleInDynamoLibrary(false)]
     public static string pCollect(string filePath, int foo, double bar);
     {
        ...
     }
}
```

And in you custom node model's `BuildOutputAst()`, change that line to:
```
[IsVisibleInDynamoLibrary(false)]
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
{
    ....        
    var funcNode = AstFactory.BuildFunctionCall(
                // Create my function MyDataCollector.pCollect with some inputs
                // Func<string, int, double, string> is my function's signature
                new Func<string, int, dobule, string>(MyDataCollector.pCollect), 
                inputAstNodes);
    ....
}
```

And that's it! If you connect `File Path` to, say the first input port of custom node, you should get a string. 

We don't directly deal with specific nodes in a custom node's implementation (for example, won't care about if the input is `CodeBlockNodeModel` or whatever else). The inputs are either primitive types like integer, string, a List of string or some public types that defined in custom node library and will be imported by Dynamo.

**Comment by jhubers** [comment]
Hi @ke-yu Thanks for your help. Sorry, for asking simple questions. I am a beginner. SampleLibraryUI.Examples.CustomNodeModel.BuildOutputAst(). I opend D:...\Dynamo-RC0.8.0_master\Dynamo-RC0.8.0_master\src\Dynamo.All.2013.sln. Not the right version?
In there I found something similar to your example: 
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
        {
            //var functionCall = AstFactory.BuildFunctionCall("Color", "BuildColorFromRange", inputAstNodes);
            var functionCall =
                AstFactory.BuildFunctionCall(
                    new Func<Color, Color, double, Color>(Color.BuildColorFromRange),
                    inputAstNodes);
            return new[]
            {
                AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), functionCall)
            };
        }
I found the definition of Func:
public delegate TResult Func<in T1, in T2, in T3, out TResult>(T1 arg1, T2 arg2, T3 arg3); comments say:  "Encapsulates a method that has three parameters and returns a value of the  type specified by the TResult parameter."
And so I understand that between () is a method that I should implement as a property of a public static class that returns the three arguments. 
In your example: 
new Func<string, int, dobule, string>(MyDataCollector.pCollect) it will return a string.  So I could write
string myString = new Func<string, int, double, string>(MyDataCollector.pCollect)
Then var functionCall = (AstFactory.BuildFunctionCall(myString, inputAstNodes) while all inputAstNodes are processed? I guess I have to make a list<AstNodes> of only inputAstNodes[index], where index is the integer of the corresponding input. And replace inputAstNodes with that list? But what parameters (string, int, double ) should be given to MyDataCollector.pCollect? You write:
public static string pCollect(string filePath, int foo, double bar);
But the filePath is what I am looking for? And what should go into foo and bar?
BTW how do you make these nice code snippets with grey background?

**Comment by ke-yu** [comment]
hi @jhubers , you could find the file here: `https://github.com/DynamoDS/Dynamo/blob/RC0.8.1_master/src/Libraries/Samples/SampleLibraryUI/Examples/CustomNodeModel.cs`.

And yes, when you write 
```var functionCall = AstFactory.BuildFunctionCall(new Func<string, int, dobule, string>(MyDataCollector.pCollect), inputAstNodes)...
```
you could expect that all `inputAstNodes` are processed (although here they aren't, but at run-time yes): these inputs finally will be converted to primitive type like int, double, string, etc. and input to your C# static function `MyDataCollector.pCollect()`, so you C# function doesn't need to worry about all Dynamo related data types, just deal with C# primitive type in most cases. 

`inputAstNodes` in `BuildOutputAst()` is already a list, no need to change it, unless you don't need some of them of want to change them...

The function `pCollect()` that I wrote here just is an example, you could change its signature based on your requirement. 

Regarding to `filePath`, when you connect the output from `File Path` node, the receiver will get a string of the full path of file that you select.

**Comment by jhubers** [comment]
@ke-yu Thanks for the link to the newer version. But still it is not clear to me. In that version I indeed see that a public static class is used. But there it returns a value:
```
public static class SampleUtilities
    {
        [IsVisibleInDynamoLibrary(false)]
        public static double MultiplyInputByNumber(double input)
        {
            return input * 42;
        }
    }
```
So, I think I have to return a string then. But if I do it like:
```
var funcNode = AstFactory.BuildFunctionCall(
                new Func<string, string>(MyDataCollector.filePath),inputAstNodes);
```
and my class:
```
 public static class MyDataCollector
    {
        [IsVisibleInDynamoLibrary(false)]
        public static string filePath(string _filePath)
        {
            return _filePath;
        }
    }
```
Nothing happens. In debugging mode it shows that var funcNode becomes
{pCOLADnamespace.MyDataCollector.filePath(var_870b02ecd0574a5585b1d8a8017e4b58_out0, var_b97e5c2766024e14b2d0be99c8518a19_out0, var_dc8ab1c4298f45a9a162f2e5f8151a49_out0, temp_0_ca98164ada7e407fb237217fd91598b6)}	ProtoCore.AST.AssociativeAST.AssociativeNode {ProtoCore.AST.AssociativeAST.IdentifierListNode}. But stepping through the code doesn't go to the MyDataCollector class. How do I get the path in a string variable so I can use it to open a file? 

Do I have to install version 0.8.1?

**Comment by holyjewsus** [comment]
hi @jhubers please try defining your DataCollector class in another assembly, this means another project than the custom node itself is defined in. I think this may help.

**Comment by jhubers** [comment]
Hi @holyjewsus  in another project? That means another dll? Have to reference it then in the first project. Tried it. There is no difference. For testing I used:
`var t = new Func<string, string>(MyDataCollector.MyDataCollector.filePath);`
When debugging t shows: 
![image](https://cloud.githubusercontent.com/assets/1529297/8410298/a380e32a-1e7d-11e5-87ad-d7eca83ae275.png)

**Comment by ke-yu** [comment]
hi @jhubers, stepping through the code won't go to `MyDataCollector` class. Here it is compiling inputs and your custom node logic to internal DesignScript representation, and later on the execution engine will hook up the static function in `MyDataCollector` class and run it. 

Setting a break point in your static function to verify whether it will be executed in the future or not.

**Comment by jhubers** [comment]
hi @ke-yu I have set a break point in MyDataGollector class. But it isn't hit. Also the node gives an error:
![image](https://cloud.githubusercontent.com/assets/1529297/8410874/2fdb65e0-1e81-11e5-9515-32c412e1f8d4.png)

**Comment by jhubers** [comment]
That error was due to not all inputs being connected. But still not clear to me, how I can get a string variable with the file path...

**Comment by holyjewsus** [comment]
Hi @jhubers 

![screen shot 2015-06-29 at 12 06 16 pm](https://cloud.githubusercontent.com/assets/508936/8412358/5e48c6ea-1e57-11e5-8fc1-0f69b440cf4d.png)

please see this modified example of the hello world custom node model sample to see how I got a string.


      [IsVisibleInDynamoLibrary(false)]
        public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
        {
           
            var funcNode = AstFactory.BuildFunctionCall(
                new Func<string,string>(SampleUtilities.StringReturnString), 
                inputAstNodes);

     
            return new[]
            {
             
                AstFactory.BuildAssignment(
                    GetAstIdentifierForOutputIndex(0), funcNode),

                AstFactory.BuildAssignment(
                    GetAstIdentifierForOutputIndex(1),
                    AstFactory.BuildDoubleNode(awesome))
            };
        }


in sampleUtilities.cs I defined a new public static method I call from inside the buildAST method

     [IsVisibleInDynamoLibrary(false)]
        public static string StringReturnString(string instring)
        {
            return instring;
        }

**Comment by holyjewsus** [comment]
It may help to see that in this build output AST method we're not actually doing any evaluation of the node, we're just creating a syntax tree of expressions that will be evaluated in the future when the node is called with some inputs.

**Comment by jhubers** [comment]
Hi @holyjewsus  Thanks for the example, but It doesn't show how and where you can get the file path in your code so you can open a file and process it in your code. It somewhere hidden in this Func. How can I get it out?

**Comment by holyjewsus** [comment]
@jhubers you can do your processing inside that func - "stringreturnstring" You can call other c# inside that func as well, this could have side effects, like taking the string and writing it to csv.  I think, @ke-yu to make sure that is safe.

**Comment by jhubers** [comment]
Hi @holyjewsus, I think I understand what you mean. However, since it is a static class that we use, how can we access what happens inside? We can not make an instance to access a property. I found another error in my code. I forgot to change the output from 
` return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), AstFactory.BuildExprList(inputAstNodes)), };` to
`return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode), };`
But now I get next error:
![image](https://cloud.githubusercontent.com/assets/1529297/8454891/e4b98d04-2000-11e5-8d5a-cf50fcca3b30.png)
Any idea what that means?

**Comment by ke-yu** [comment]
Hi @jhubers, that means some inputs are incorrect. Could you show the signature of C# static function that `pSHARE` that invokes? What's type of input 'N'? And are 'I', 'L', 'U' strings?

**Comment by jhubers** [comment]
Hi @ke-yu, I cleaned out all extra properties and methods of pSHARE.cs. But still the error as shown in picture below. I don't see that somewhere the type of the inputs is defined. Am I missing something? N is output of List.Create node. I and L are outputs of File Path node and U is output of a Code Block node. 

The cleaned code of pSHARE.cs:
```
using System.Collections.Generic;
using Autodesk.DesignScript.Runtime;
using Dynamo.Controls;
using Dynamo.Models;
using ProtoCore.AST.AssociativeAST;
using System;
using Dynamo.Wpf;

namespace pCOLADnamespace
{
    #region some node settings
    /// pSHARE takes care of the communication of parameter changes through a shared *.csv file.
    [NodeName("pSHARE")]
    [NodeCategory("pCOLAD")]
    [NodeDescription("Load and share changes to parameters.")]
    [IsDesignScriptCompatible]
    #endregion
    public class pSHARE : NodeModel
    {

        #region constructor
        /// The constructor for a NodeModel is used to create the input and output ports and specify the argument lacing.
        public pSHARE()
        {
            InPortData.Add(new PortData("N", "Input (a List.CreateList) of pCOLLECT output(s)"));
            InPortData.Add(new PortData("I", "Input a FilePath for the shared csv files."));
            InPortData.Add(new PortData("L", "Input a FilePath for the local copy of the csv file."));
            InPortData.Add(new PortData("U", "Input a the user namen (Code Block)."));
            OutPortData.Add(new PortData("O", "Output of parameter name and value on next line; two by two."));
            RegisterAllPorts();
            ArgumentLacing = LacingStrategy.CrossProduct;
        }

        #endregion

        #region public methods
        /// <summary>
        /// If this method is not overriden, Dynamo will, by default
        /// pass data through this node. So we should later use it to pass the OutputList.
        /// </summary>
        /// <param name="inputAstNodes"></param>
        /// <returns></return
        [IsVisibleInDynamoLibrary(false)]
        public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
        {
            var t = new Func<string, string>(MyDataCollector.MyDataCollector.filePath);
            var funcNode = AstFactory.BuildFunctionCall(t,inputAstNodes);
            return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)};
        }
        public class pSHARENodeViewCustomization : INodeViewCustomization<pSHARE>
        {
            /// <summary>
            /// At run-time, this method is called during the node 
            /// creation. Here you can create custom UI elements and
            /// add them to the node view, but we recommend designing
            /// your UI declaratively using xaml, and binding it to
            /// properties on this node as the DataContext.
            /// </summary>
            /// <param name="model">The NodeModel representing the node's core logic.</param>
            /// <param name="nodeView">The NodeView representing the node in the graph.</param>

            public void CustomizeView(pSHARE model, NodeView nodeView)
            {
                var pSHAREControl = new pSHAREcontrol();
                nodeView.inputGrid.Children.Add(pSHAREControl);
                pSHAREControl.DataContext = model;
            }
            /// <summary>
            /// Here you can do any cleanup you require if you've assigned callbacks for particular 
            /// UI events on your node.
            /// </summary>
            public void Dispose() { }

        }
        #endregion
    }
}
```
The code of a separate project MyDataCollector:
```
using Autodesk.DesignScript.Runtime;

namespace MyDataCollector
{
    public static class MyDataCollector
    {
        [IsVisibleInDynamoLibrary(false)]
        public static string filePath(string _filePath)
        {         
            return _filePath;
        }
    }
}
```
![image](https://cloud.githubusercontent.com/assets/1529297/8472507/28bcddca-20a2-11e5-837a-e1e273dd8ecf.png)

**Comment by jhubers** [comment]
Might it be that the error comes from the fact that MyDataCollectorClass is an extension method. For this to work  we must put a using MyDataCollector (the name space). And a method should have () behind it? However this gives next error:
![image](https://cloud.githubusercontent.com/assets/1529297/8474772/e5bf13f6-20b2-11e5-8612-fdd6003698db.png)

Intellisense shows:
![image](https://cloud.githubusercontent.com/assets/1529297/8474853/9ffc7ee8-20b3-11e5-88a5-8089cb48c745.png)

I don't understand what should go between () in .filePath().

**Comment by ke-yu** [comment]
hi @jhubers , your custom node `pSHARE` has *4* inputs, but the C# function it tries to call, `MyDataCollectorClass.filePath(string)` only has *one parameter*, which is expected to be a string, therefore at runtime Dynamo is not able to call this static function. 

And when you are creating a `Func<>`, `()` is not needed, otherwise it will become a function call instead of creating a `Func<>` object which points to that static function.

**Comment by jhubers** [comment]
hi @ke-yu , thanks, this makes sense. But I tried now with next solution:
```
var t = new Func<List<string>,string,string, string, List<object>>(MyDataCollector.MyDataCollectorClass.pSHAREinputs);
```

And the function:
```
using Autodesk.DesignScript.Runtime;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;

namespace MyDataCollector
{
    public static class MyDataCollectorClass
    {
        [IsVisibleInDynamoLibrary(false)]
        public static List<object> pSHAREinputs( List<string> _Ninputs, string _IfilePath, string _LfilePath, string _owner)
        {
            System.Windows.MessageBox.Show("The file path = " + _IfilePath);
            List<object> _pSHAREinputs = new List<object> { _Ninputs, _IfilePath, _LfilePath, _owner };
            return _pSHAREinputs;
        }
    }
}
```
But same error of Dereferencing a non-pointer. (3f47aacd)... What else is wrong?

**Comment by ke-yu** [comment]
@jhubers , firstly, could you remove 
`[IsVisibleInDynamoLibrary(false)]` before `pSHAREInputs`, and import the dll into Dynamo to see if Dynamo is able to create a node `pSHAREInputs` and verify if it works as expected?

If yes, I don't have any clue now, could you share your code somewhere, for example on github, so that I can take a look?

**Comment by jhubers** [comment]
@ke-yu  Thanks again for the help. See result in first image. I had to load the dll manually though to see it in the Dynamo menu. So just putting it in the nodes folder of Dynamo didn't work. In the meanwhile I installed version 0.8.1 and had to change `Inputs` into `InputNodes` at some places. But I don't think that interfered with getting the inputs for pSHARE. I noticed that I expect to get a `List<string>` from pCOLLECT, while it shows in the image that it is` string[]`. Will change that. Otherwise when I run the solution I get an other error message now (see second image): Warning: Converting an array to var would cause array rank reduction and is not permitted.
![image](https://cloud.githubusercontent.com/assets/1529297/8497244/eed90d24-217b-11e5-915c-68a6088a96f9.png)
![image](https://cloud.githubusercontent.com/assets/1529297/8497250/f6a520e2-217b-11e5-884f-e790801de343.png)

I would like to put my code on GitHub, but I don't know how. Will try to find out. Next week I don't have time to work on this. But next two weeks I will. Now first try with `string[]`. Will give the result in some minutes.

**Comment by ke-yu** [comment]
@jhubers , `List<String>` is the type in C#, but `string[]` is the type in DesignScript. That said, as the data will flow from C# to Dynamo or Dynamo to C#, internally Dynamo will do data marshalling, so no need to worry about `List<String>` and `string[]`.

**Comment by jhubers** [comment]
@ke-yu Indeed that didn't change. So remain two problems. MyDataCollector.dll doesn't load automatically. If I load it manually. The a warning shows: Converting an array to var would cause array rank reduction and is not permitted. That that mean that I should change `var t = new Func...` into an array. But of what type? 
Anyway I have a correct output now and also it breaks inside MyDataCollector.cs and shows that I can use the inputs. Thanks again for your help. Will now try to get everything on GitHub.

**Comment by jhubers** [comment]
@ke-yu I think the repository is ready: https://github.com/jhubers/pCOLAD-for-Dynamo. Please tell me if it usable like this...

**Comment by jhubers** [comment]
It turns out that the warning about Converting an array to var comes from the List<string> input.  If I change it to string there is no problem. In fact it is not the List<string> input, but the fact that it is added to the List<object> as such. If I convert the List<string> to a string with returns and then add it to the List<object>, there is no problem. Anyway the output should be a List<string> and that works alright.

**Comment by jhubers** [comment]
Remains the problem of the MyDataCollector.dll not loading automatically. What is needed for a dll to load automatically when it is in the nodes directory of Dynamo?

**Comment by ke-yu** [comment]
@jhubers, try to put libraries in `nodes` folder (not verified yet).

**Comment by jhubers** [comment]
@ke-yu Of course. MyDataCollector.dll is in the nodes folder. Just like the pCOLAD.dll. I copy them there with a post-build event. Verified that they are of same recent date. Is it because it is a `public static class MyDataCollectorClass` and a `public static List<string> pSHAREinputs(List<string> _Ninputs, string _IfilePath, string _LfilePath, string _owner)`?

**Comment by ke-yu** [comment]
hmm, how about creating a [package](https://github.com/DynamoDS/Dynamo/wiki/Building-a-Package-for-Dynamo-in-Visual-Studio)?

**Comment by jhubers** [comment]
I tried to reorganize my GitHub repository several times, but found that folder structures in VS are not relevant to GitHub. Reorganizing the folder structure in Explorer caused all kinds of errors in VS. Had to revert, but created the package in a different place. But when the package is in the roaming folder of Dynamo 0.8 next to a downloaded package, only the last one loads. Is the `pkg.json` file necessary? How do you make that?

**Comment by ke-yu** [comment]
@pboyer , @riteshchandawar , any tool to create `pkg.json`?

**Comment by jhubers** [comment]
@ke-yu  I made a pkg.json file on the basis of the Zip List package. It seems to work. However I get next error. I made a new repository: https://github.com/jhubers/pCOLAD-for-Dynamo
![image](https://cloud.githubusercontent.com/assets/1529297/8648116/4632346c-295e-11e5-9b58-b08da07770df.png)

**Comment by jhubers** [comment]
@ke-yu I made pkg.json file on the basis of the Zip List package. It works alright now. Thanks for your help.

**Comment by jhubers** [comment]
Now I can access the inputs in this `public static class MyDataCollector`. But I need that information in the `pSHARE node` in order to add it to the csv file and its display in xaml DataGrid. However the `public static class MyDataCollector` is only called just before creation of the node. If I try to get it after `return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)};` a warning shows that this code is not reachable. I can not reference `pSHARE` in `MyDataCollector`, because then a circular reference occurs. So, how can I use the inputs?

**Comment by ke-yu** [comment]
hi @jhubers , you can't reference `pSHARE`. What kind of information that can't be passed to `MyDataCollector`?

It is a little bit tricky, but you can always create some AST nodes for some values, if they are not from input, and pass them as parameters to you static function defined in `MyDataCollector`. 

For example, in your `BuiltOutAst()`:

```
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
 {
        var t = new Func<string, string>(MyDataCollector.MyDataCollector.filePath);

    // instead of using inputASTNodes,
    // you can create new input. e.g.,
    inputAstNodes[1] = AstFactory.BuildString("foo");

    var funcNode = AstFactory.BuildFunctionCall(t,inputAstNodes);
    return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)};
}
```

**Comment by jhubers** [comment]
@ke-yu the problem was that in the static class I can not use `RaisePropertyChanged`, which I needed to react on when clicking on a checkbox and probably need it also to show if something changes in the input nodes. Therefore moving all the code to static class `MyDataCollector` didn't work. Now I have it mixed. Everything that interacts with the display is in the `pSHARE` node and the rest in static classes. Seems to work, but not yet updating changes in the inputs, without closing and opening the display xaml control. But I am a step further. Thanks.

**Comment by jhubers** [comment]
Hi @ke-yu I can't get the updating of the `CSVControl` to work when I change a `pCOLLECT` input and hit the RUN button of Dynamo. Seems a problem is that using a static class you cannot implement the `INotifyPropertyChanged` interface, because that needs an instance. The problem is that we then can not `RaisePropertyChanged("MyPropDataTable")` to notify the `CSVcontrol` that it should update changes form `myDataTable`. E.g. after changing inputs of `pSHARE` or changes in the csv file. The way to go would be to use a singleton? But I don't know if that would work with the 
`Func<List<List<string>>, string, string, string, List<string>>(MyDataCollectorClass.pSHAREinputs).`
What do you think?

The repository is now at https://github.com/jhubers/pCOLAD-for-Dynamo/tree/0.1

**Comment by ke-yu** [comment]
hi @jhubers, I'll take a look when i have time.

**Comment by ke-yu** [comment]
hi @jhubers , how about event handler? Can we add a event handler to MyDataCollector and add CSVControl as a subscriber and once MyDataCollector finishes execution, it notifies CSVControl to update?

**Comment by jhubers** [comment]
@ke-yu Indeed. An event in the static class MyDataCollectorClass does the trick. Still many things to improve, but the coming 4 weeks I am on holiday. Thank you very much for your help. Without it I would have given up.

**Comment by ke-yu** [comment]
@jhubers , hope all these difficulties won't stop you from exploring and hacking Dynamo! :-) At the meantime, we (Dynamo team) should think about how to make it easy for other developers to extend the functionalities of Dynamo!

**Comment by jhubers** [comment]
Hi @ke-yu Back from holidays, picking it up again. Noticed that my pCOLLECT script gives an error when using automatic run in Dynamo. Anyway it didn't have the right method for getting the inputs. But I was wondering if there is a solution to have dynamic inputs. I mean that you can add or delete inputs like with the List.Create node, using + or -? But how can we then use the BuildFunctionCall, when we don't know what inputs there are?

**Comment by ke-yu** [comment]
@jhubers , to do that, you need a c# function which accepts any input, e.g.,
```
class MyClass
{
    public static int Sum(params int[] vs)
    {
        int sum = 0;
        foreach (int v in vs)
            sum += v;
        return sum;
    }
}
```
And create a variable input UI node, which could derive from `VariableInputNode`

**Comment by jhubers** [comment]
Hi @ke-yu. Thanks. The difficulty for me is that I have this example code to create the output of my nodes: 
```
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
        {
            var t = new Func<List<List<string>>, string, string, string, List<string>>(MyDataCollectorClass.pSHAREinputs);
            var funcNode = AstFactory.BuildFunctionCall(t, inputAstNodes);
            return new[] 
            { 
                AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)
             };
        }
```
I don't know how to replace the new Func<…> by a c# function which accepts any input. I guess BuildFunctionCall is expecting a functionName as first parameter then. But what should this static function return? If I return a string, Dynamo displays an error:
![image](https://cloud.githubusercontent.com/assets/1529297/9577333/0e39fa56-4fe0-11e5-81f0-00555ff1b247.png)

**Comment by ke-yu** [comment]
@jhubers , you don't need to use `Func<...>` stuff, it just helps to automatically figure out its class name and function name (`AstFactory.BuildFunctionCall()` has a bunch of overloaded version, the basic two are `AssociativeNode BuildFunctionCall(string className, string functionName, List<AssociativeNode> arguments, Core core = null)` and `AssociativeNode BuildFunctionCall(string functionName, List<AssociativeNode> arguments, Core core = null)`. You could explicitly do that:
```
var funcNode = AstFactory.BuildFunctionCall("MyClass", "pCollect", inputAstNodes);
```

**Comment by ke-yu** [comment]
`className` is fully qualified name.

**Comment by jhubers** [comment]
Hi @ke-yu That is what I did. The function returned a string ("P;V;I;C;O;A;"). Which are the nick names of the inputs. That is a mistake of mine, but why do I get an error stating: "Warning: Method 'P;V;I;C;O;A;()' not found"? This was the code:
```
 List<string> pCOLLECTinputsList = new List<string>();
            foreach (PortData item in InPortData)
            {
                pCOLLECTinputsList.Add(item.NickName);
            }

            string[] pCOLLECTinputs = pCOLLECTinputsList.ToArray();
            var funcNode = AstFactory.BuildFunctionCall(MyDataCollector.MyDataCollectorClass.pCOLLECToutputs(pCOLLECTinputs), inputAstNodes);
            return new[] 
            { 
                AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)
             };
```
And the function:
```
 public static string pCOLLECToutputs(params string[] ss)
        {
            //pCOLLECT should output a list of ;-separated strings in the format:
            //Parameter;New Value;Importance;Comments;Owner;Extra Attribute Name;Extra Attribute Name; ...etc
            string pCOLLECToutput = "";
            foreach (string s in ss)
            {
                pCOLLECToutput += s;
                pCOLLECToutput += ";";
            }
            return pCOLLECToutput;
        }
```

**Comment by ke-yu** [comment]
hi @jhubers , try
```
AstFactory.BuildFunctionCall("MyDataCollector.MyDataCollectorClass", "pCOLLECToutputs", ...)
```

**Comment by ke-yu** [comment]
Or 
```
AstFactory.BuildFunctionCall("MyDataCollector.MyDataCollectorClass.pCOLLECToutputs", ...)
```

**Comment by jhubers** [comment]
Nope, still errors:
```
Warning: Cannot find static method or constructor MyDataCollector.MyDataCollectorClass.pCOLLECToutputs()
```
or 
```
Warning: Method 'MyDataCollector.MyDataCollectorClass.pCOLLECToutputs()' not found
```

**Comment by mjkkirschner** [comment]
@jhubers the dataCollector class is defined in another project / assembly and referenced by this project?

**Comment by jhubers** [comment]
@mjkkirschner Yes it is: using MyDataCollector; is at the top of the code. Other functions in it work ok.

**Comment by mjkkirschner** [comment]
hmm, I'm not sure, but it seems like you should not be passing the the inputASTnodes and the your input string to the function where the buildFunctionCall method is expecting a function name, not a function with the arguments set... 

I think using another overload where an action is specified is the correct way, but I'm not sure which one.

**Comment by jhubers** [comment]
@mjkkirschner  Well the definition of the BuildFunctionCall is:
```
        public static AssociativeNode BuildFunctionCall(
            string functionName, List<AssociativeNode> arguments, Core core = null)
        {
            if (string.IsNullOrEmpty(functionName))
                throw new ArgumentException("functionName");

            if (arguments == null)
                throw new ArgumentNullException("arguments");

            var funcCall = new FunctionCallNode
            {
                Function = BuildIdentifier(functionName),
                FormalArguments = arguments
            };
            return funcCall;
        }
```
So I guess you have to pass the function name as string, the arguments as List<AssociativeNode>. But the Core core = null part I don't understand. But if I add null as parameter, nothing changes.

**Comment by jhubers** [comment]
Hi @ke-yu, Any other things I can try?

**Comment by ke-yu** [comment]
hi @jhubers , let's try again...load the library that contains pCollect firstly to see if Dynamo could find this class or not.

**Comment by jhubers** [comment]
Hi @ke-yu. Dynamo can find the pCOLLECT class. How else could it display the node? I am using a tool that attaches the process to Dynamo. So I can debug the behavior. I can see that Dynamo goes through the code and even makes an AssociateNode... But then displays the error that it can not find the function. This function is in an other assembly. That assembly loads too, because it is used in the pSHARE and pPARAM node, and works fine.

**Comment by ke-yu** [comment]
So if it could be loaded, then you are able to use in code block node, right? Currently I have no clue why function couldn't be found. :-(

**Comment by jhubers** [comment]
@ke-yu. I am not sure what you mean with "use this function in code block node"? If I double click on the canvas and type p it doesn't show pCOLAD in the list of possible commands. But in the menu at the left there is a pCOLAD entry. If I type pCOLAD.pCOLLECT in the code block and hit enter it displays the warning:
`Warning: Internal error, please report: Dereferencing a non-pointer. (3f47aacd)`

**Comment by jhubers** [comment]
@ke-yu. When I type MyCollectorDataClass.pCOLLECToutputs, then I have next result
![image](https://cloud.githubusercontent.com/assets/1529297/9683405/2cad1468-5310-11e5-8521-24d7bd88f52b.png)

**Comment by ke-yu** [comment]
hi @jhubers , can you try `MyDataCollector.MyDataCollectorClass.pCOLLECToutputs()` to see if it generates some result? If yes, then we should be able to compile a node to this one. Not quite sure where goes wrong. :-(

**Comment by jhubers** [comment]
@ke-yu Now it gives: `Warning: One or more of the input types are not matching. Couldn't find a version of pCOLLECToutputs that takes arguments of type (var)`
![image](https://cloud.githubusercontent.com/assets/1529297/9712091/7b65e3ca-5548-11e5-9404-6ff856aff8a0.png)
I was using:
```
        public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
        {
            List<string> pCOLLECTinputsList = new List<string>();
            foreach (PortData item in InPortData)
            {
                pCOLLECTinputsList.Add(item.NickName);
            }

            string[] pCOLLECTinputs = pCOLLECTinputsList.ToArray();
            var funcNode = AstFactory.BuildFunctionCall("MyDataCollector.MyDataCollectorClass.pCOLLECToutputs", inputAstNodes);
            return new[] 
            { 
                AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), funcNode)
             };
        }
```
And in MyDataCollector.MyDataCollectorClass:
```
        public static string pCOLLECToutputs(params string[] ss)
        {
            string pCOLLECToutput = "";
            foreach (string s in ss)
            {
                pCOLLECToutput += s;
                pCOLLECToutput += ";";
            }
            return pCOLLECToutput;
        }

```

**Comment by ke-yu** [comment]
@jhubers , is there a overloaded `pCOLLECToutputs()`? The warning says there are mutiple version of `pCOLLECToutputs()`.

**Comment by jhubers** [comment]
@ke-yu Not that I know. A search also shows only one not commented out `pCOLLECToutputs()`
```
Find all "pCOLLECToutputs", Find Results 1, Entire Solution, ""
  D:\Data\Research\Dynamo\pCOLAD for Dynamo\pCOLLECT2.cs(256):            var funcNode = AstFactory.BuildFunctionCall("MyDataCollector.MyDataCollectorClass.pCOLLECToutputs", inputAstNodes);
  D:\Data\Research\Dynamo\pCOLAD for Dynamo\pCOLLECT2.cs(298):        //    //delS newDelS = new delS(MyDataCollector.MyDataCollectorClass.pCOLLECToutputs);
  D:\Data\Research\Dynamo\pCOLAD for Dynamo\pCOLLECT2.cs(313):        //    var funcNode = AstFactory.BuildFunctionCall("MyDataCollector.MyDataCollectorClass.pCOLLECToutputs(inputNames)", inputAstNodes);
  D:\Data\Research\Dynamo\pCOLAD for Dynamo\MyDataCollector\MyDataCollector.cs(144):        //public static List<string> pCOLLECToutputs(params string[] ss)
  D:\Data\Research\Dynamo\pCOLAD for Dynamo\MyDataCollector\MyDataCollector.cs(165):        public static string pCOLLECToutputs(params string[] ss)
  Matching lines: 5    Matching files: 2    Total files searched: 19
```

**Comment by jhubers** [comment]
@ke-yu Can you check the `List.Reduce` node? I don't quite understand what it does, but I have the impression it is not functioning correctly either. So maybe something wrong with the `VariableInputNode`?

**Comment by ke-yu** [comment]
@jhubers , could you upload code to github so that I could debug it when I get some free time?

**Comment by jhubers** [comment]
@ke-yu. Thanks. You can find the repository here: https://github.com/jhubers/pCOLAD-for-Dynamo/tree/0.1

**Comment by ke-yu** [comment]
hi @jhubers , I can't find `pCOLLECToutputs()` function, where it is defined and used?

**Comment by jhubers** [comment]
Hi Yu Ke,

You can find it at the bottum of: https://github.com/jhubers/pCOLAD-for-Dynamo/blob/0.1/MyDataCollector/MyDataCollector.cs

Best regards,

Dr. Ir. J.C. (Hans) Hubers
Associate professor

TU Delft / Faculty of Architecture
Julianalaan 134 / 2628 BL, Delft
01+.West.030
PO Box 5043 / 2600 GA, Delft
The Netherlands

**Comment by jhubers** [comment]
It is used in public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes) which you can find here:
https://github.com/jhubers/pCOLAD-for-Dynamo/blob/0.1/pCOLLECT2.cs

Best regards,

Dr. Ir. J.C. (Hans) Hubers
Associate professor

TU Delft / Faculty of Architecture
Julianalaan 134 / 2628 BL, Delft
01+.West.030
PO Box 5043 / 2600 GA, Delft
The Netherlands

**Comment by jhubers** [comment]
Be aware that you must download the 0.1 branch on GitHub…

Best regards,

Dr. Ir. J.C. (Hans) Hubers
Associate professor

TU Delft / Faculty of Architecture
Julianalaan 134 / 2628 BL, Delft
01+.West.030
PO Box 5043 / 2600 GA, Delft
The Netherlands

**Comment by ke-yu** [comment]
hi @jhubers , try 

```
var exprList = AstFactory.BuildExprList(inputAstNodes);
var funcNode = AstFactory.BuildFunctionCall("MyDataCollector.MyDataCollectorClass", "pCOLLECToutputs", new List<AssociativeNode> { exprList });
```

**Comment by jhubers** [comment]
hi @ke-yu It gives next result
![image](https://cloud.githubusercontent.com/assets/1529297/10043185/32f289ee-61f1-11e5-8f9b-cd82b1c357be.png)
So pCOLLECT gives warning: Warning: Asked to convert non-convertible types
MyDataCollectorClass.pCOLLECToutputs operation failed. And the code block: Warning: One or more of the input types are not matching. Couldn't find a version of pCOLLECToutputs that takes arguments of type (var).

**Comment by jhubers** [comment]
@ke-yu Remember I was using:
```
        public static string pCOLLECToutputs(params string[] ss)
        {
            //pCOLLECT should output a list of ;-separated strings in the format:
            //Parameter;New Value;Importance;Comments;Owner;Extra Attribute Name;Extra Attribute Name; ...etc
            string pCOLLECToutput = "";
            foreach (string s in ss)
            {
                pCOLLECToutput += s;
                pCOLLECToutput += ";";
            }
            return pCOLLECToutput;
        }
```

**Comment by jhubers** [comment]
@ke-yu Hi, It's been a while. I continued with pSHARE and solved the display in red of changes since last time the csv file was saved. Now I am comming back to the variable input component pCOLLECT again. I am using:
```
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
    //build a new output List<AssociativeNode>consisting of fieldnames seperated by ';' and
    // on next inputAstNodes with ';' added
    List<AssociativeNode> pCOLLECTtempList = new List<AssociativeNode>();
    string inputNames = "";
    foreach (var item in InPortData)
    {
        inputNames += item.ToolTipString + ";";
    }
    inputNames = inputNames.Remove(inputNames.Length - 1);
    var headings = AstFactory.BuildStringNode(inputNames);
    var semiColon = AstFactory.BuildStringNode(";");
    foreach (AssociativeNode InputItem in inputAstNodes)
    {
        List<AssociativeNode> arguments = new List<AssociativeNode>();
        arguments.Add(InputItem);
        arguments.Add(semiColon);
        var funcNode = AstFactory.BuildFunctionCall("%add", arguments);
        //don't add ';' to the last one
        if (inputAstNodes.IndexOf(InputItem) == inputAstNodes.Count - 1)
        {
            pCOLLECTtempList.Add(InputItem);
        }
        else
        {
            pCOLLECTtempList.Add(funcNode);
        }
    }
    // now pCOLLECTtempList has the inputs followed by ';'
    // but it should become one string so add the items together
    List<AssociativeNode> pCOLLECToutputList = new List<AssociativeNode>();
    AssociativeNode A = pCOLLECTtempList[0];
    for (int i = 0; i < pCOLLECTtempList.Count - 1; i++)
    {
        List<AssociativeNode> arguments = new List<AssociativeNode>();
        arguments.Add(A);
        arguments.Add(pCOLLECTtempList[i + 1]);
        var funcNode = AstFactory.BuildFunctionCall("%add", arguments);
        A = funcNode;
    }
    pCOLLECToutputList.Add(A);
    pCOLLECToutputList.Insert(0, headings);

    return new[]
    {
        AstFactory.BuildAssignment(
            GetAstIdentifierForOutputIndex(0), AstFactory.BuildExprList(pCOLLECToutputList)),
    };
}
```
But now next problem. When you add an input and give it a name and tooltip, the next time you start your Dynamo solution the name and tooltip are replaced by "index[index]" and "No tooltip available". That is not possible. I use the names as column name in the csv file. Is there a way to store the input names and tooltips for every pCOLLECT? Best would be of course in the *.dyn file...

**Comment by ke-yu** [comment]
hi @jhubers , though you assign tooltip in `BuildOutputAst()`, I don't think tool tip will be saved to .dyn file. 

For variable input node (base is `VariableInputNode`), typically we override its `GetInputTooltip(int index)` to return a customized tooltip at some index. An example is `List.Create` node (path: src\Libraries\CoreNodeModels\CreateList.cs), it overrides `GetInputTooltip()`. 

But in your case, as the tooltip is from some column in csv, it is dynamic, it needs to be saved to .dyn file. To achieve that, overriding two functions:
```
protected override void SerializeCore(XmlElement element, SaveContext context)
{
    base.SerializeCore(element, context);
    // save customized tooltip information to .dyn file, which is a xml file
    ....
}

protected override void DeserializeCore(XmlElement nodeElement, SaveContext context)
{
    base.DeserializeCore(nodeElement, context);
    // read customized tooltip information from .dyn file, which is a xml file
    ....
}
```

You may take a look at `NodeModel.SerializeCore()` to see how to save information to .xml (src\DynamoCore\Models\NodeModel.cs)

After you do that, save .dyn and open it with any text editor to verify xml element is properly added.

**Comment by jhubers** [comment]
@ke-yu Great, Thanks. It works. Had to change heritance from VariableInputNode to NodeModel though. Otherwise it would also add the extra inputs as index[index].

**Comment by jhubers** [comment]
Next thing. The user should get a warning when somebody changes the csv file. That works ok with a systemfilewatcher. But I would like to detect if automatic run is true. If so then I would like to update the solution and use the new information. If run is manual, then I would like to change the state of the solution so that when the user hits the Run button  the last information in the csv file will be used. So two questions:
1) How can I detect if solution is running manual or automatic.
2) How can I set the state of the solution so that the Run button uses the last information.

**Comment by ke-yu** [comment]
hi @jhubers , I don't think csv reading/writing node implements file watcher. One way to do that is to implement in a customized node by using `FileSystemWatcher` (e.g., https://msdn.microsoft.com/en-us/library/system.io.filesystemwatcher(v=vs.110).aspx). Then once the node detects the file is modified, call `OnNodeModified(forceExecute: true)`. 

Pseudo code looks like:
```
class FileMonitorNode: NodeModel, ...
{
       public void OnFileModified(...)
       {
            this.OnNodeModified(forceExecute:true);
        }
}
```

**Comment by jhubers** [comment]
@ke-yu Thanks. And is there a way to detect if the solution is running manual or automatic?

**Comment by ke-yu** [comment]
`HomeWorkspaceModel.RunSettings.RunType`

**Comment by sharadkjaiswal** [comment]
If you use File Path->File.FromPath->CSV.ReadFromFile node to select the file to read and connect it to csv file reader, then it will automatically support FileWatcher feature.
![image](https://cloud.githubusercontent.com/assets/1754137/11426184/21e0b320-9495-11e5-9365-5ee7f8023562.png)

**Comment by jhubers** [comment]
@ke-yu I get an error: An object reference is required for the non-static field, method, or property 'Dynamo.Models.HomeWorkspaceModel.RunSettings'.

**Comment by ke-yu** [comment]
@jhubers, try to get the instance of `DynamoModel`, and from it you could get a list of `WorkSpaces`, and get the workspace model whose type is `HomeWorkspaceModel`...

**Comment by jhubers** [comment]
@ke-yu The only method I found to get the instance of DynamoModel is 
`DynamoModel dm = DynamoModel.Start(new DynamoModel.DefaultStartConfiguration());` 
or 
`DynamoModel dm = DynamoModel.Start();`
Both always show that the RunType is Automatic. While it actually is Manual.

**Comment by mjkkirschner** [comment]
you can get an instance of a DynamoModel by getting the view model from the node view, you can do this inside the UI customization<type> for your node.

**Comment by jhubers** [comment]
@mjkkirschner I can get the viewmodel, indeed inside the UI customization through the NodeView parameter: `Dynamo.ViewModels.NodeViewModel vm = nodeView.ViewModel;` But for some reason intellisense doesn't work on vm and `vm.Model` does not exist.

**Comment by jhubers** [comment]
@mjkkirschner Think I got it: 
```
Dynamo.ViewModels.NodeViewModel vm = nodeView.ViewModel;
Dynamo.Models.NodeModel nm = vm.NodeModel;
Dynamo.ViewModels.DynamoViewModel dvm = vm.DynamoViewModel;
DynamoModel dm = dvm.Model;
```

**Comment by jhubers** [comment]
Had to reinstall Windows 8.1. Installed VS 2015 and Dynamo 9.0. Several references didn't load of course. I resolved them all. Also went to Prism 6. My solution builds fine now and the dll's etc. are copied to the ``C:\Users\jhubers\AppData\Roaming\Dynamo\0.9\packages\pCOLADpackage\bin`` folder. However my components don't show up anymore in the menu and in the test dyn file they are stamped 'Unresolved'. What can be wrong?

**Comment by jhubers** [comment]
Maybe found the problem. I forgot about this pkg.json file again. I've made it on the basis of the Reshuffel Lists Package pkg.json file. Now pCOLAD packages appears next to Reshuffel Lists Package in the Package Manager. However both don't show up in the menu. The Mangae Node and Package Path is set to ``C:\Users\jhubers\AppData\Roaming\Dynamo\0.9`` Please help...
![image](https://cloud.githubusercontent.com/assets/1529297/11567728/a7dc12fa-99e8-11e5-910f-069398747d2a.png)

**Comment by jhubers** [comment]
In the DynamoSamples.2013.sln, in the CustomNodeModel.cs still is stated that "The assembly which contains this node needs to be located in the 'nodes' folder in Dynamo in order to be loaded at startup." But is this true in the case of a package? Also in the pkg.json file should the setting "contains_binaries":true"? And maybe something should go into node_libraries":[]?

**Comment by jhubers** [comment]
Indeed when I copy the dll's to the node folder, the pCOLAD menu shows up. The nodes in the test.dyn file show up fine. Only get an error again in pSHARE: ``Warning: Internal error, please report: Dereferencing a non-pointer. (3f47aacd).`` How can I debug this? With attached process it is not detected...

**Comment by mjkkirschner** [comment]
hi @jhubers have you looked in the dynamo console to see what exception if any there is during load?

have you seen this document? https://github.com/DynamoDS/Dynamo/wiki/API-Changes

**Comment by mjkkirschner** [comment]
node libraries should contain any .dlls what will be converted to nodes, contains binaries should be true if the package contains any binaries I believe.

**Comment by mjkkirschner** [comment]
I don't think that is the case with packages - but packages must be in the package folder to load UI nodes correctly.

**Comment by jhubers** [comment]
Just putting the names of the dlls in the pkg.json like so: 
```
{"license":"","file_hash":null,"name":"pCOLAD","version":"0.0.1","description":"Tools for on-line sharing of parameters.","group":"pCOLAD","keywords":["pCOLAD","share"],"dependencies":[],"contents":"pCOLAD - Tools for on-line sharing of parameters","engine_version":"0.9.1.3067","engine":"dynamo","engine_metadata":"","site_url":"","repository_url":"","contains_binaries":true,"node_libraries":["MyDataCollector.dll","pCOLAD.dll"]}
```
Doesn't work.

**Comment by jhubers** [comment]
@ke-yu ``var funcNode = AstFactory.BuildFunctionCall(function, inputAstNodes);``doesn't work anymore in Dynamo 9.0. Is this a known bug?

**Comment by jhubers** [comment]
There was something wrong in the pkg.json file. When I adapted it from the DynamoSamples-0.9.0 files everything works ok again.

**Comment by jhubers** [comment]
Start testing with Dynamo in Revit. Run into problem with `Application.Current.Windows` Which I use to close eventual WPF controls that are left open when closing Dynamo. `Apllication.Current` is null in Revit. Is there another trick to make sure that I don't open a WPF control if there is one already?

**Comment by ke-yu** [comment]
@Benglin, @Randy-Ma  any idea?

**Comment by jhubers** [comment]
Can I detect when Dynamo is closing? I could then close my controls too...

**Comment by mjkkirschner** [comment]
@jhubers if you implementing a Dynamo UI node from Nodemodel than you can access the events thrown by the DynamoView or the DynamoModel and subscribe to them, if its a zero touch node, look in DynamoServices for the events we have made public.

**Comment by jhubers** [comment]
@mjkkirschner  Yes I am implementing a Dynamo UI node. But how can I access the events of DynamoView or DynamoModel. Should I override the `OnShutdownStarted`? If so, where do I do that? In the derived Nodemodel class, or in the constructor? Nothing seems to work...

**Comment by mjkkirschner** [comment]
Hi Hans,

You can subscribe a method on your node model  to the events of the dynamo model 

https://msdn.microsoft.com/en-us/library/ms366768.aspx

First you need to get a reference to the dynamo model.

**Comment by mjkkirschner** [comment]
If you want to use the onshutdownstsrted event you would do the same thing - just subscribe to it in your constructor and subscribe some new method that will be called when the event fires - I believe this method is called when Dynamo model is shutting down but I'm not 100% that it's called. I can't check at this moment.

**Comment by jhubers** [comment]
Still struggling with this. I know how to get a reference to `DynamoModel` through a nested class that is implementing `INodeViewCustomization`. But when I try to get to a method of the parent class by passing the parent to the constructor of the nested class, the customization of the node doesn't work any more. Is there a way to get the DynamoModel at the start of the constructor of the node?

**Comment by jhubers** [comment]
Found one reason: I didn't know that if you pass the parent to an instance constructor, you also need a default constructor. So now I can access my custom node's methods. But when closing Dynamo the Dispose method is of no use, because the custom node is already null. So have to find a method before the custom node closes...

**Comment by ke-yu** [comment]
Hi @jhubers , is really necessary to know if Dynamo is shutting down in node? A node's life cycle is shorter than DynamoModel, even shorter than workspace, so ideally, a node should not know anything about DynamoModel

**Comment by ke-yu** [comment]
@jhubers , if you really need DynamoModel, in your class (inherits from INodeViewCustomization), the constructor should have a second parameter `NodeView`, and from that you can get DynamoModel: `nodeView.ViewModel.DynamoViewModel.Model`, and probably you are interested in `DynamoModel.ShutdownStarted` event...

**Comment by jhubers** [comment]
@ke-yu Well for the prototype I guess it is al-right to leave the display of csv files behind when you close Dynamo. You can close it separately. But in a professional plug-in it should of course not be possible. And indeed the event `DynamoModel.ShuwtdownStarted`is not received. Probably because the node is closed before. So maybe in future it will be necessary to send an event before closing the nodes...

**Comment by jhubers** [comment]
Trying to keep up with Dynamo updates, now with version 0.9.1.4062 I get a warning: `'NodeModel.InPortData' is obsolete: 'InPortData is deprecated, please use the InPortNamesAttribute, InPortDescriptionsAttribute, and InPortTypesAttribute instead.'` But how do I use these attributes? E.g.
`InPortNamesAttribute ipna = new InPortNamesAttribute(new string[] { "N", "I", "L", "U" });` doesn't give an error but the input doesn't appear neither.

**Comment by jhubers** [comment]
Updated to Dynamo 0.9.1.4062, I have several errors. Some references had to be update. Now while debugging I get `error CS1704: An assembly with the same simple name 'ProtoCore' has already been imported. Try removing one of the references (e.g. 'ProtoCore.dll') or sign them to enable side-by-side.` But I checked several times; ProtoCore is only once in the References.
Also `InportData` is obsolete and several attributes should be used. But how can I add inputs dynamically then and serialize and deserialize them?

**Comment by ke-yu** [comment]
@jhubers do you have different version of Protocore.dll?

**Comment by jhubers** [comment]
@ke-yu Yes many different versions found on C-drive. See picture below. Very strange also in the packages folder. This was maybe due to having copy local set to true in the properties of ProtoCore reference. But if I set it to false and rebuild the solution it reappears like in other picture.
![image](https://cloud.githubusercontent.com/assets/1529297/13253299/af8b944c-da3b-11e5-8d07-32dcf643ca4b.png)
![image](https://cloud.githubusercontent.com/assets/1529297/13253319/cb9d9a5e-da3b-11e5-943f-2f7bd4c00629.png)

And yes adding input dynamically like List.Create with the plus sign.

**Comment by jhubers** [comment]
@ke-yu Yes many different versions found on C-drive (see picture below). I removed the highlited one from the Dynamo packages and also from the bin folder where it was copied from. But after rebuilding and restarting VS2015 still the same problem. The error prevents me from debugging. In Diagnostic I also see errors like: `Exception thrown: 'ProtoCore.DSASM.Mirror.SymbolNotFoundException' in ProtoCore.dll ("Cannot find symbol: var_9d81e05418af490abc9912b702d3fb87")`. But that is probably related to the same problem...
 
![image](https://cloud.githubusercontent.com/assets/1529297/13257022/621b6f7c-da4d-11e5-99d4-de347c28c220.png)

And yes adding input dynamically like List.Create with the plus sign.

**Comment by jhubers** [comment]
@ke-yu The problem with importing ProtoCore twice is solved. I removed Dynamo and made a clean install. Remains the problem of using attributes for dynamically adding inports.

**Comment by jhubers** [comment]
Solved the problem of leaving controls behind in separate issue:[https://github.com/DynamoDS/Dynamo/issues/6231](url)

**Comment by jhubers** [comment]
In the left menu of Dynamo 1.0.0 I have several undesirable entries (CSharpAnalytics, Cyotek, DynamoCoreWpf, FontAwesome, GraphLayout, LibG, MyDataCollector, ProtoScript, RestSharp). Probably has something to do with missing `[IsVisibleInDynamoLibrary(false)]` attributes. But where should they be put?
![image](https://cloud.githubusercontent.com/assets/1529297/15091332/ba0d288c-1446-11e6-93e0-c3200b71e477.png)

**Comment by jhubers** [comment]
`InportData` is obsolete and several attributes should be used. But how can I add inputs dynamically then and serialize and deserialize them?

**Comment by ke-yu** [comment]
@jhubers `InportData` is still used internally. Let's check with related developer to confirm whether this obsolete attribute is still appropriate or not.

**Comment by jhubers** [comment]
@ke-yu Hi, please do check the use of attributes instead off the `InportData` property (`ObservableCollection<PortData>`) in the case of a node with variable number of inputs. Like e.g. the `CreateList` node. I don't understand how that could work. Here ([https://github.com/DynamoDS/Dynamo/pull/5340](url)) I found a discussion concerning this issue, but I don't fully understand it. When I check the definition of `CreateList` indeed `InportData` is still used. So how can it be obsolete then? Could you provide some code how to add input ports on the fly and how to serialize and deserialize them without using the obsolete `InportData`?

**Comment by ke-yu** [comment]
@ramramps , couple of input port and output port related properties and methods are marked as obsolete, e.g., `NodeModel.RegisterOutputPorts()`, `NodeModel.InPortData` and so on. What are replacements for these methods and properties?

**Comment by jhubers** [comment]
@ke-yu Hi, I tried several things, but couldn't get it to work. When I use the `InPorts` with a UI node that  can add inputs, the first Run the `InPorts` are fine, based on the attributes. But when I try to add a `PortModel`,  `InPorts` is empty. Solved that by storing the `InPorts` in a property as follows:
```
List<PortModel> lpm = InPorts.ToList();
myInports = new ObservableCollection<PortModel>(lpm);
```
That works, and I can add `PortModel` to `myInports`. But when I try then to fill the `InPorts` as follows:
```
this.myInports.Add(new PortModel(PortType.Input, this, new PortData(a1, a2)));
List<PortModel> lmpm = myInports.ToList();
InPorts = new ObservableCollection<PortModel>(lmpm);
RegisterAllPorts();
```
All input ports are gone. RegisterAllPorts() apparently empties InPorts. If I put `InPorts = new ObservableCollection<PortModel>(lmpm);` again after RegisterAllPorts(); the old input ports are there and an empty place for the new input port is made, but no connector.

**Comment by ke-yu** [comment]
hi @jhubers , I think you might try `InPortData` instead of `InPorts`. E.g., 
```
InPortData.Clear();
InPortData.Add(new PortData(...));
RegisterInputPorts();
```

**Comment by jhubers** [comment]
@ke-yu `InPortData` is obsolete. That is what I try to stop using...
